challenges/backend/src/app/services/CarOnSaleClient/classes/CarOnSale.ts

```diff
@@ -0,0 +1,113 @@
+
+        this.printNumberOfAuctions(auctionResponses);
+        this.printAverageBids(auctionResponses);
+        this.printAveragePercentage(auctionResponses);
+    }
+
+    async getRunningAuctions(token: string, filter: IActionFilter, count: boolean): Promise<Auction> {
+        const url = `${this.config.get('api_endpoint')}v2/auction/buyer/?${JSON.stringify(filter)}&count=${count}`;
+        const response = await this.axiosInstance.get(url, {
+            headers: {
+                'userid': this.config.get('email'),
+                'authtoken': token,
+                'Content-Type': 'application/json',
+                'Accept': 'application/json',
+            }
+        });
+
+        if (!response) {
+            throw new Error("invalid input");
+        }
+
+        const auction: Auction = {
+            items: response.data.items,
+            page: response.data.page,
+            total: response.data.total,
+        };
+
+        return auction;
+    }
+    async userAuthentication(): Promise<AuthenticationResponse> {
+        try {
+            const password = this.config.get('password');
+            const meta = this.config.get('meta');
+
+            const response = await this.axiosInstance.put(`${this.config.get('api_endpoint')}v1/authentication/${this.config.get('email')}`, {
+                password,
+                meta,
+            }, {
+                headers: {
+                    'Content-Type': 'application/json',
+                    'Accept': 'application/json',
+                }
+            });
+            if (!response.data) {
+                throw new Error("invalid credentials");
+
+            }
+            return response.data;
+
+        } catch (error: any) {
+            this.logger.log(error);
+            throw new Error(error)
+        }
+    }
+
+    private printNumberOfAuctions(auctions: AuctionItem[]): void {
+        this.logger.log(`Number of Auctions: ${auctions.length}`);
+    }
+
+    private printAveragePercentage(auctions: AuctionItem[]): void {
+        const averageProgress = this.calculator.calculateAverageProgress(auctions);
+        this.logger.log(`Average auction progress: ${averageProgress.toFixed(2)}%`);
+    }
+    private printAverageBids(auctions: AuctionItem[]): void {
+        const averageBids = this.calculator.calculateAverageBids(auctions);
+        this.logger.log(`Average average bid: ${averageBids}`);
+    }
+    private async prepareAuctionData(numPages: number, token: string, page_limit: number): Promise<AuctionItem[]> {
+        const requests = Array.from({ length: numPages }, (_, i) => this.getRunningAuctions(token, { limit: page_limit, offset: i * page_limit }, false));
+
+        const responses: Auction[] = await Promise.all(requests);
+        const items: AuctionItem[] = responses.flatMap(response => response.items);
+        return Array.from(new Set(items));
+    }
```

Comment:
Nice move! You warned yourself to search all available pages for the calculation.